fili-core/src/main/java/com/yahoo/bard/webservice/data/DruidQueryBuilder.java

```diff
@@ -252,9 +254,9 @@ protected GroupByQuery buildGroupByQuery(
      *
      * @return A table datasource for a fact table or a union data source for a fact table view
      */
-    private DataSource buildTableDataSource(PhysicalTable table) {
-        if (table instanceof ConcretePhysicalTable) {
-            return new TableDataSource((ConcretePhysicalTable) table);
+    private DataSource buildTableDataSource(ConstrainedTable table) {
+        if (table.getDataSourceNames().size() == 1) {
+            return new TableDataSource(table, table.getDataSourceNames().stream().findFirst().get());
         } else {
             return new UnionDataSource(Sets.newHashSet(table));
```

Comment:
We probably don't want to make everything Union, on the off-chance that Druid has an optimization around the single-table version vs. the union version. That said, I can definitely see this optimization being moved out of the main query builder and into a static analysis step at some point.